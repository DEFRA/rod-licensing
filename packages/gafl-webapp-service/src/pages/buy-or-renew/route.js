import { BUY_OR_RENEW } from '../../uri.js'
import Joi from 'joi'
import pageRoute from '../../routes/page-route.js'
import { nextPage } from '../../routes/next-page.js'
export const buyOrRenew = {
  buy: 'buy',
  renew: 'renew'
}

export const validator = Joi.object({
  'buy-or-renew': Joi.string()
    .valid('buy-licence', 'renew-licence')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

const route = pageRoute(BUY_OR_RENEW.page, BUY_OR_RENEW.uri, validator, nextPage)

export default route
