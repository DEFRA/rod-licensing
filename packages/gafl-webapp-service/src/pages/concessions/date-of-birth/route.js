import { DATE_OF_BIRTH } from '../../../uri.js'
import Joi from 'joi'
import pageRoute from '../../../routes/page-route.js'
import { validation } from '@defra-fish/business-rules-lib'
import { nextPage } from '../../../routes/next-page.js'

export const validator = payload => {
  const dateOfBirth = `${payload['date-of-birth-year']}-${payload['date-of-birth-month']}-${payload['date-of-birth-day']}`
  Joi.assert(
    { 'date-of-birth': dateOfBirth },
    Joi.object({
      'date-of-birth': validation.contact.createBirthDateValidator(Joi)
    })
  )
}

export const getData = async request => {
  const { isLicenceForYou } = await request.cache().helpers.status.getCurrentPermission()

  return { isLicenceForYou }
}

export default pageRoute(DATE_OF_BIRTH.page, DATE_OF_BIRTH.uri, validator, nextPage, getData)
