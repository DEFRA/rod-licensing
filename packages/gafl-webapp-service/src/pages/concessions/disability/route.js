import { DISABILITY_CONCESSION } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import * as concessionHelper from '../../../processors/concession-helper.js'
import { nextPage } from '../../../routes/next-page.js'
import { disabilityConcessionTypes } from './update-transaction.js'

const validator = Joi.object({
  'disability-concession': Joi.string()
    .valid(...Object.values(disabilityConcessionTypes))
    .required(),
  'ni-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.pipDla,
    then: validation.contact.createNationalInsuranceNumberValidator(Joi).required(),
    otherwise: Joi.string().empty('')
  }),
  'blue-badge-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.blueBadge,
    then: Joi.string()
      .max(25)
      .required(),
    otherwise: Joi.string().empty('')
  })
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    hasJunior: concessionHelper.hasJunior(permission),
    hasSenior: concessionHelper.hasSenior(permission),
    isLicenceForYou: permission.isLicenceForYou,
    ...disabilityConcessionTypes
  }
}

export default pageRoute(DISABILITY_CONCESSION.page, DISABILITY_CONCESSION.uri, validator, nextPage, getData)
