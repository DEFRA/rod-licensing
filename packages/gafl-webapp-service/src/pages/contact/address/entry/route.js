import { ADDRESS_ENTRY, ADDRESS_LOOKUP } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import { countries } from '../../../../processors/refdata-helper.js'
import { nextPage } from '../../../../routes/next-page.js'
import { getPronoun } from '../../../../processors/licence-type-display.js'

export const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  street: validation.contact.createStreetValidator(Joi),
  locality: validation.contact.createLocalityValidator(Joi),
  town: validation.contact.createTownValidator(Joi),
  postcode: Joi.alternatives().conditional('country-code', {
    is: Joi.string().pattern(/^GB/),
    then: validation.contact.createUKPostcodeValidator(Joi),
    otherwise: validation.contact.createOverseasPostcodeValidator(Joi)
  }),
  'country-code': Joi.string().required()
}).options({ abortEarly: false, allowUnknown: true })

export const getCountryDropDownOptions = async () => {
  const options = await countries.getAll()
  return options.filter(country => country.code !== 'GB')
}

export const getData = async request => {
  const { addresses, searchTerms } = await request.cache().helpers.addressLookup.getCurrentPermission()
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()

  const pronoun = getPronoun(isLicenceForYou).possessive

  return {
    pronoun,
    searchTerms: !addresses?.length && searchTerms ? searchTerms : null,
    countries: await getCountryDropDownOptions(),
    lookupPage: ADDRESS_LOOKUP.uri
  }
}

export default pageRoute(ADDRESS_ENTRY.page, ADDRESS_ENTRY.uri, validator, nextPage, getData)
