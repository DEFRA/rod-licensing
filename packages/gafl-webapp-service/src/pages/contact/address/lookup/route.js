import { ADDRESS_LOOKUP, ADDRESS_ENTRY, OS_TERMS } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { hasJunior, validation } from '@defra-fish/business-rules-lib'
import { isPhysical } from '../../../../processors/licence-type-display.js'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  postcode: validation.contact.createUKPostcodeValidator(Joi)
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    isLicenceForYou: permission.isLicenceForYou,
    licenceLength: permission.licenceLength,
    junior: hasJunior(permission),
    isPhysical: isPhysical(permission),
    uri: {
      entryPage: addLanguageCodeToUri(request, ADDRESS_ENTRY.uri),
      osTerms: addLanguageCodeToUri(request, OS_TERMS.uri)
    }
  }
}

export default pageRoute(ADDRESS_LOOKUP.page, ADDRESS_LOOKUP.uri, validator, nextPage, getData)
