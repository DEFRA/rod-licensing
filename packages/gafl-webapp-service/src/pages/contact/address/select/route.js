import { ADDRESS_SELECT, ADDRESS_LOOKUP, ADDRESS_ENTRY } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

export const getData = async request => {
  const { addresses, searchTerms } = await request.cache().helpers.addressLookup.getCurrentPermission()
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    isLicenceForYou,
    addresses,
    searchTerms,
    lookupPage: addLanguageCodeToUri(request, ADDRESS_LOOKUP.uri),
    entryPage: addLanguageCodeToUri(request, ADDRESS_ENTRY.uri)
  }
}

export default pageRoute(
  ADDRESS_SELECT.page,
  ADDRESS_SELECT.uri,
  Joi.object({
    address: Joi.number().integer().min(0).required()
  }).options({ abortEarly: false, allowUnknown: true }),
  nextPage,
  getData
)
