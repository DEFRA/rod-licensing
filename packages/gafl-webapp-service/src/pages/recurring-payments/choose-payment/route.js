import { CHOOSE_PAYMENT } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { nextPage } from '../../../routes/next-page.js'

export const validator = Joi.object({
  'recurring-payment': Joi.string().valid('yes', 'no').required()
}).options({ abortEarly: false, allowUnknown: true })

export default pageRoute(CHOOSE_PAYMENT.page, CHOOSE_PAYMENT.uri, validator, nextPage)
