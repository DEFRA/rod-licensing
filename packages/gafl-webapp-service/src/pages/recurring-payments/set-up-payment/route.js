import { CHOOSE_PAYMENT, SET_UP_PAYMENT, TERMS_AND_CONDITIONS } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { nextPage } from '../../../routes/next-page.js'
import { recurringLicenceTypeDisplay } from '../../../processors/licence-type-display.js'
import { addLanguageCodeToUri } from '../../../processors/uri-helper.js'
import { recurringPayReminderDisplay } from '../../../processors/recurring-pay-reminder-display.js'
import { displayPermissionPrice } from '../../../processors/price-display.js'

export const validator = Joi.object({
  agree: Joi.string().valid('yes').required()
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    cost: displayPermissionPrice(permission, request.i18n.getCatalog()),
    type: recurringLicenceTypeDisplay(permission, request.i18n.getCatalog()),
    reminder: recurringPayReminderDisplay(permission, request.i18n.getCatalog()),
    uri: {
      single: addLanguageCodeToUri(request, CHOOSE_PAYMENT.uri),
      terms: addLanguageCodeToUri(request, TERMS_AND_CONDITIONS.uri) // update when rp terms and conditions created
    }
  }
}

export default pageRoute(SET_UP_PAYMENT.page, SET_UP_PAYMENT.uri, validator, nextPage, getData)
