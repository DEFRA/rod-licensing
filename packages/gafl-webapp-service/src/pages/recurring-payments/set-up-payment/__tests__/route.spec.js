import { validator, getData } from '../route.js'
import { nextPage } from '../../../../routes/next-page.js'
import pageRoute from '../../../../routes/page-route.js'
import { recurringLicenceTypeDisplay } from '../../../../processors/licence-type-display.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'
import { CHOOSE_PAYMENT, SET_UP_PAYMENT, TERMS_AND_CONDITIONS } from '../../../../uri.js'
import { recurringPayReminderDisplay } from '../../../../processors/recurring-pay-reminder-display.js'
import { displayPermissionPrice } from '../../../../processors/price-display.js'

jest.mock('../../../../routes/page-route.js')
jest.mock('../../../../processors/licence-type-display.js')
jest.mock('../../../../processors/uri-helper.js')
jest.mock('../../../../processors/recurring-pay-reminder-display.js')
jest.mock('../../../../processors/price-display.js')

const getSampleRequest = (permission = {}, catalog = getCatalog()) => ({
  cache: () => ({
    helpers: {
      transaction: {
        getCurrentPermission: () => permission
      }
    }
  }),
  i18n: {
    getCatalog: () => catalog
  }
})

const getCatalog = () => ({
  recurring_payment_set_up_error: 'recurring payment error'
})

describe('route', () => {
  describe('validator', () => {
    it('validator should validate "yes" as a valid choice', () => {
      const result = validator.validate({ agree: 'yes' })
      expect(result.error).toBeUndefined()
    })

    it('validator should not validate an invalid choice', () => {
      const result = validator.validate({ agree: 'invalid' })
      expect(result.error).toBeDefined()
    })
  })

  describe('default', () => {
    it('should call the pageRoute with set-up-payment, /buy/set-up-recurring-card-payment, validator, nextPage and getData', async () => {
      expect(pageRoute).toBeCalledWith(SET_UP_PAYMENT.page, SET_UP_PAYMENT.uri, validator, nextPage, getData)
    })
  })

  describe('getData', () => {
    it('returns expected values', async () => {
      const cost = Symbol('cost value')
      const type = Symbol('type value')
      const reminder = Symbol('reminder value')
      const uri = Symbol('uri value')

      displayPermissionPrice.mockReturnValueOnce(cost)
      recurringLicenceTypeDisplay.mockReturnValueOnce(type)
      recurringPayReminderDisplay.mockReturnValueOnce(reminder)
      addLanguageCodeToUri.mockReturnValue(uri)

      const result = await getData(getSampleRequest())

      expect(result).toMatchSnapshot()
    })

    it.each([
      [displayPermissionPrice, Symbol('mock catalog'), Symbol('mock permission')],
      [recurringLicenceTypeDisplay, Symbol('mock catalog'), Symbol('mock permission')],
      [recurringPayReminderDisplay, Symbol('mock catalog'), Symbol('mock permission')]
    ])('%s is called with permission and getCatalog', async (func, catalog, permission) => {
      const request = getSampleRequest(permission, catalog)

      await getData(request)

      expect(func).toHaveBeenCalledWith(permission, catalog)
    })

    it.each([[CHOOSE_PAYMENT.uri], [TERMS_AND_CONDITIONS.uri]
    ])('addLanguageCodeToUri is called with permission and %s', async (uri) => {
      const request = getSampleRequest()

      await getData(request)

      expect(addLanguageCodeToUri).toHaveBeenCalledWith(request, uri)
    })
  })
})
