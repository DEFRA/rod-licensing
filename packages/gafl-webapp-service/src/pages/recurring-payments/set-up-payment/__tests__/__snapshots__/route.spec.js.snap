// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route getData returns expected values 1`] = `
Object {
  "cost": Symbol(cost value),
  "errorMap": Object {
    "agree": Object {
      "any.required": Object {
        "ref": "#agree",
        "text": "recurring payment error",
      },
    },
  },
  "reminder": Symbol(reminder value),
  "type": Symbol(type value),
  "uri": Object {
    "single": Symbol(uri value),
    "terms": Symbol(uri value),
  },
}
`;
